// js-csp/lib/impl/instruction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AltsInstruction = exports.SleepInstruction = exports.PutInstruction = exports.TakeInstruction = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _channels = require('./channels');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TakeInstruction = exports.TakeInstruction = function TakeInstruction(channel) {
  (0, _classCallCheck3.default)(this, TakeInstruction);

  this.channel = channel;
};

var PutInstruction = exports.PutInstruction = function PutInstruction(channel, value) {
  (0, _classCallCheck3.default)(this, PutInstruction);

  this.channel = channel;
  this.value = value;
};

var SleepInstruction = exports.SleepInstruction = function SleepInstruction(msec) {
  (0, _classCallCheck3.default)(this, SleepInstruction);

  this.msec = msec;
};

var AltsInstruction = exports.AltsInstruction = function AltsInstruction(operations, options) {
  (0, _classCallCheck3.default)(this, AltsInstruction);

  this.operations = operations;
  this.options = options;
};
